src/workerd/api/queue.c++

```diff
@@ -47,32 +125,36 @@ kj::Promise<void> WorkerQueue::send(
   }).attach(kj::mv(client));
 };
 
+struct SerializedBody {
+  kj::Array<kj::byte> body;
+  kj::Maybe<kj::StringPtr> contentType;
+};
+
 kj::Promise<void> WorkerQueue::sendBatch(
-    jsg::Sequence<MessageSendRequest> batch, v8::Isolate* isolate) {
+    jsg::Lock& js, jsg::Sequence<MessageSendRequest> batch) {
   auto& context = IoContext::current();
 
   JSG_REQUIRE(batch.size() > 0, TypeError, "sendBatch() requires at least one message");
 
   size_t totalSize = 0;
   size_t largestMessage = 0;
   auto messageCount = batch.size();
-  auto builder = kj::heapArrayBuilder<kj::Array<byte>>(messageCount);
+  auto builder = kj::heapArrayBuilder<SerializedBody>(messageCount);
   for (auto& message: batch) {
-    JSG_REQUIRE(!message.body.getHandle(isolate)->IsUndefined(), TypeError,
+    JSG_REQUIRE(!message.body.getHandle(js)->IsUndefined(), TypeError,
         "Message body cannot be undefined");
 
-    // Use a specific serialization version to avoid sending messages using a new version before all
-    // runtimes at the edge know how to read it.
-    // TODO(perf): Would we be better off just serializing all the messages together in one big
-    // buffer rather than separately?
-    jsg::Serializer serializer(isolate, jsg::Serializer::Options {
-      .version = 15,
-      .omitHeader = false,
-    });
-    serializer.write(kj::mv(message.body));
-    builder.add(serializer.release().data);
-    totalSize += builder.back().size();
-    largestMessage = kj::max(largestMessage, builder.back().size());
+    SerializedBody item;
+    KJ_IF_MAYBE(contentType, message.contentType) {
+      item.contentType = validateContentType(*contentType);
+      item.body = serialize(js, message.body.getHandle(js), *contentType);
+    } else {
+      item.body = serializeV8(js, message.body.getHandle(js));
+    }
+
+    builder.add(kj::mv(item));
+    totalSize += builder.back().body.size() + builder.back().contentType.orDefault("").size();
+    largestMessage = kj::max(largestMessage, builder.back().body.size());
```

Comment:
I'm thinking it might be best to remove this from the size computation.

It looks like the broker is already slightly broken, in that it will accept a message payload of 128KB, but will then add additional fields like a timestamp and ID before storing a message, which could then make the "StoredMessage" over the max DO value size, right?

Plus, I'm thinking I'll go ahead and optimize the broker side to use something closer to an enum, so the size computation here wouldn't even exactly match what we put into storage.

Reply:
> over the max DO value size, right?

No, because we're cheating -- the DO max value size is 128 KiB whereas the queue max value size is 128 KB.

But that doesn't mean that there's nothing potentially broken once we start letting users set custom metadata of their own.

Reply:
Ah, well, that's fun 😆, a little bit of extra space at least.

I actually went ahead and removed it from the size calculation.